js/components/contextMenu.js

```diff
@@ -257,54 +253,149 @@ class ContextMenu extends ImmutableComponent {
   }
 
   onKeyDown (e) {
-    const currentIndex = this.currentIndex
-
-    if (e.keyCode === KeyCodes.ESC || e.keyCode === KeyCodes.TAB) {
-      windowActions.setContextMenuDetail()
-    } else if (e.keyCode === KeyCodes.ENTER) {
-      e.preventDefault()
-      const action = this.props.contextMenuDetail.get('template').get(currentIndex).get('click')
-      if (action) {
-        action(e)
-      }
-    } else if (e.keyCode === KeyCodes.DOWN && currentIndex < this.maxIndex) {
-      windowActions.setSubmenuSelectedIndex([0, currentIndex + 1])
-    } else if (e.keyCode === KeyCodes.UP && currentIndex > 0) {
-      windowActions.setSubmenuSelectedIndex([0, currentIndex - 1])
+    const selectedIndex = (this.props.selectedIndex === null) ? [0] : this.props.selectedIndex
+    const currentIndex = selectedIndex[selectedIndex.length - 1]
+    const selectedTemplate = this.getMenuByIndex(selectedIndex, this.props.contextMenuDetail.get('template'))
+    const selectedMenuItem = selectedTemplate.get(currentIndex)
+
+    switch (e.keyCode) {
+      case keyCodes.ENTER:
+        e.preventDefault()
+        e.stopPropagation()
+        const action = selectedTemplate.getIn([currentIndex, 'click'])
+        if (action) {
+          action(e)
+        }
+        windowActions.resetMenuState()
+        break
+
+      case KeyCodes.ESC:
+      case KeyCodes.TAB:
+        windowActions.resetMenuState()
+        break
+
+      case keyCodes.LEFT:
+        // Left arrow inside a sub menu
+        // <= go back one level
+        if (e.which === keyCodes.LEFT && this.hasSubmenuSelection) {
+          const newIndices = selectedIndex.slice()
+          newIndices.pop()
+          windowActions.setContextMenuSelectedIndex(newIndices)
+
+          let openedSubmenuDetails = this.props.contextMenuDetail.get('openedSubmenuDetails')
+            ? this.props.contextMenuDetail.get('openedSubmenuDetails')
+            : new Immutable.List()
+          openedSubmenuDetails = openedSubmenuDetails.pop()
+
+          windowActions.setContextMenuDetail(this.props.contextMenuDetail.set('openedSubmenuDetails', openedSubmenuDetails))
+        }
+        break
+
+      case keyCodes.RIGHT:
+        // Right arrow on a menu item which has a sub menu
+        // => go up one level (default next menu to item 0)
+        if (e.which === keyCodes.RIGHT && selectedMenuItem.has('submenu')) {
```

Comment:
We should check `!!selectedMenuItem.get('submenu')` here. You can press right at the end of submenu to see what happened. It will be `undefined` on popup menu and `null` on menubar